### PR TITLE
Provide more options to search bar users

### LINKED ISSUE #9: Need a way to focus the search bar
Use case - Need to show keyboard automatically,  when moved to the search bar. 
Current implementation - Now, we need to focus the search text field manually. 
Expected - Need to focus the text filed while moving to the search bar.

autofocus: true does the job on the TextField (Search bar). 

It could be useful if you expose the TextFiled properties or a toggle callback for autofocus thing.
Thanks in advance.

### LINKED ISSUE #11: Not able clear the last searched text on SearchBar
I am facing issue in clearing the search text. I have a requirement to clear the previously searched text when user wants to search for the second time. Could you please help me to achieve this functionality?

### LINKED ISSUE #12: How to set height of the search bar
I need to set search bar height, how can I achieve it?

### LINKED ISSUE #13: Request - Add a child at the left side of the search bar.
A small request.
Will like to add a back button at the left side of search bar.
Hope you could add a leftchild parameter for this.
Thanks 😃

### LINKED ISSUE #14: Disable auto search
Feature request: disable the auto search and only search when the user presses return.

### LINKED ISSUE #15: Autofocus feature
Hey there,

I am implementing the Flappy Search Bar as a new screen in my, which appears when a user clicks on a search button somewhere else in my app. Therefore it would be quite useful if autofocus would be supported as it is in the Flutter TextField class for example. As for now the user needs to click in the Searchbar again to open the keyboard. I havent found a property which looks like the one I need.

Thank you very much!
Daniel

### PR DESCRIPTION
**Pull request by kylerwsm**
# Overview of Changes
- Moved `SearchBarStyle` into `flappy_search_bar.dart`. This enables users of the search bar to use `SearchBar` while defining their own `SearchBarStyle` without the need for an additional import statement.
- Added new options in `SearchBarStyle`
  - `surroundingColor`: Color of the search bar area, including the optional leading and trailing widgets
  - `searchBarHeight`: Allows users to set the height of the search bar
    - Closes #12
- Added new options in `SearchBar`
  - `centerEmptyWidget`: Whether the empty widget should be centered if shown
  - `useCancellationWidget`: Indicates whether the user wants to use the cancellation widget
  - `enableSuggestions`: Whether the keyboard should show dictionary suggestions
  - `searchOnlyOnSubmit`: Whether search should only have when the searched button is pressed.
    - The done button on the keyboard has been changed to search.
    - Closes #14
  - `contentPadding`: The padding on the searched result
  - `leading`: Allow users to use a leading widget
    - Closes #13 
  - `trailing`: Allow users to use a trailing widget
  - `autoFocus`: Whether the search bar should get focus on first launch
    - Closes #15 
  - `focusNode`: Allows users to use pass in their own `FocusNode`
    - Closes #9
  - `textEditingController`: Allows users to pass in their own `TextEditingController`
    - Closes #11

# Misc Changes
- Upgraded the `dart sdk` min version to `2.3.0` to use `if` statements in widgets.

### CODE REVIEW

**Comment by mkfeuhrer** [comment]
@kylerwsm When are you planning to merge this?

**Comment by kylerwsm** [comment]
@mkfeuhrer I cannot merge this as I am not a maintainer of this repo. Am looking forward to get this merged too!

**Comment by mkfeuhrer** [comment]
@ThomasEcalle Can you please review and merge this? This would be of quite some help. Thanks :)

File: lib/flappy_search_bar.dart
Comment (dariof28) [inline lib/flappy_search_bar.dart]: Could you remove this hard coded value too please? If the user want to expand the Container for some reason, he could use `widget.searchBarPadding`
Is there any reason why this value has been hard-coded?

Comment (kylerwsm) [reply lib/flappy_search_bar.dart]: Hi, which hard-coded value are you referring to? If you are referring to `height: 80`, that was previously there. What I have done in this part was to only wrap the entire widget with a `Container`, allowing users to add a background color to the area surrounding the search bar.

Comment (dariof28) [reply lib/flappy_search_bar.dart]: Yes I'm talking about the `height: 80`. I know it was already there, but since you are editing this part I was wondering if you could remove it.
This fixed value lets the `Container` that wrap the `SearchBar` having a fixed height and is not nice in some layout situations.

Comment (dariof28) [reply lib/flappy_search_bar.dart]: The surroundingArea that will be filled with `surroundingColor` could be rendered giving a `widget.searchBarPadding`. This way if no `Padding` value is provided the real bar element (the `Row`) will take the whole `Container` and the total height will be `widget.searchBarStyle.searchBarHeight`.

Right now even with zero `Padding` and even if I set a `widget.searchBarStyle.searchBarHeight` of 40.0 for example, I'll have the `Row` taking 80.0 on height

Comment (kylerwsm) [reply lib/flappy_search_bar.dart]: Will spend some time to do this in a few days time. Will appreciate any help from anyone to edit in the mean time!

**Comment by jamesdixon** [comment]
@kylerwsm nice work! hope to see this merged soon 👍

File: lib/flappy_search_bar.dart
Comment (kylerwsm) [reply lib/flappy_search_bar.dart]: @dariof28 changes are done. Do you have write access in this repo?

**Comment by kylerwsm** [comment]
Thanks @jamesdixon, I am still waiting on @ThomasEcalle or anyone with write access to review and merge this PR.

File: lib/flappy_search_bar.dart
Comment (ThomasEcalle) [inline lib/flappy_search_bar.dart:168]: I am not sure about the utility of that boolean
If the user wants a centered empty widget, he just has to give a centered widget like : Center(child: MyWidget)

This way, the user is free to center (or not) the widget and I think it is more the Flutter's way to compose widgets like this.

File: lib/flappy_search_bar.dart
Comment (ThomasEcalle) [inline lib/flappy_search_bar.dart:192]: This is interesting as users should effectively be able to hide it.
But I think that we should hide the SearchBar animation also if we hide the cancellation widget, what do you think ?

File: lib/flappy_search_bar.dart
Comment (ThomasEcalle) [inline lib/flappy_search_bar.dart:250]: Enabling autoFocus on the `example` app throws an error : 

```
flutter: ══╡ EXCEPTION CAUGHT BY FOUNDATION LIBRARY ╞════════════════════════════════════════════════════════
flutter: The following assertion was thrown while dispatching notifications for FocusNode:
flutter: RenderBox was not laid out: RenderEditable#60a72 NEEDS-LAYOUT NEEDS-PAINT
flutter: 'package:flutter/src/rendering/box.dart':
flutter: Failed assertion: line 1687 pos 12: 'hasSize'
flutter:
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new?template=BUG.md
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #2      RenderBox.size (package:flutter/src/rendering/box.dart:1687:12)
flutter: #3      EditableTextState._updateSizeAndTransform (package:flutter/src/widgets/editable_text.dart:1729:40)
flutter: #4      EditableTextState._openInputConnection (package:flutter/src/widgets/editable_text.dart:1415:7)
flutter: #5      EditableTextState._openOrCloseInputConnectionIfNeeded (package:flutter/src/widgets/editable_text.dart:1441:7)
flutter: #6      EditableTextState._handleFocusChanged (package:flutter/src/widgets/editable_text.dart:1707:5)
flutter: #7      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:206:21)
flutter: #8      FocusNode._notify (package:flutter/src/widgets/focus_manager.dart:808:5)
flutter: #9      FocusManager._applyFocusChange (package:flutter/src/widgets/focus_manager.dart:1401:12)
flutter: (elided 12 frames from class _AssertionError and package dart:async)
```

File: lib/flappy_search_bar.dart
Comment (ThomasEcalle) [inline lib/flappy_search_bar.dart:256]: Why dont you use the Search Bar Controller as an interface to edit the TextField ?
Using a "Controller" in a package should enable the user to control anything, don't you agree ?

File: lib/flappy_search_bar.dart
Comment (ThomasEcalle) [inline lib/flappy_search_bar.dart:457]: I do not understand the idea of the `autoFocus` logic 
It seems here that you want it to launch a first search, then focus the search_bar.
But : 
* In my point of view, auto focusing a search bar should not launch a search, just put the cursor in the bar
* As it is the first launch, `_searchQueryController.text` is empty
* Why don't you simply use the `autoFocus` field on TextField ?

File: lib/flappy_search_bar.dart
Comment (ThomasEcalle) [inline lib/flappy_search_bar.dart:496]: I like the idea of implementing a `textInputAction`  but why couldn't we give the opportunity to the user to change it itself ?

**Review by ThomasEcalle — CHANGES_REQUESTED** [review]
Hi @kylerwsm  !

First of all, I am sorry for the delay.
The situation in my company (in France) with the Covid-19 has forced us to put our open source projects on the side for a moment...  But I am back ! :)

I thank you a lot for your work !
I appreciate the idea of this PR but I have several questions :

* About the boolean to center the "empty widget" :  why couldn't we let the users center or not their widget ? I think that it is more the Flutter's way as it enable the user to inject exactly the Widget he wants

* About the `useCancellationWidget` : Nice idea, but should't we also hide the animation on the search bar if we hide the cancellation widget ? Also, it seems that the search bar now translates a little to the top when a search is launched, quite weird

* About the `autoFocus` : I do not understand the logic in the code
At first it seems that enabling autoFocus throws an error (I have pasted it in the review)
About the auto focus itself, it seems that you launch a search at first build.. but is it really this logic that a user is waiting for when setting `autoFocus` to true ? For me autoFocus would simply be the value given to the TextField `autoFocus` fiels, no ?
As it is the first launch, the search that you try to make seems useless.

Do you see what I mean ? Not sure to be clear enough 


Thanks a lot for your work anyway !
I'll try to be more  present on this package, despite the difficult context that had lead us to change a lot the way we work here :)

File: lib/flappy_search_bar.dart
Comment (dariof28) [reply lib/flappy_search_bar.dart]: Unfortunately no :pensive:

**Comment by lucasjinreal** [comment]
Does this merges?

**Comment by emretuna** [comment]
Appriciate the work you guys doing, hope autofocus works.

**Comment by lucasjinreal** [comment]
So does it merged or not?

**Comment by dariof28** [comment]
Any news here?

File: lib/flappy_search_bar.dart
Comment (kylerwsm) [reply lib/flappy_search_bar.dart:496]: At this point since this PR is already sufficiently huge, I believe abstracting the `textInputAction` for user customisation may be appropriate for a separate PR.

**Comment by kylerwsm** [comment]
Sorry for the late reply. Things have gotten really busy for me recently. Will try to do up as much as possible over the weekends. If someone can help me with some of the issues pointed out by @ThomasEcalle, I will really appreciate it 🙏

**Comment by daveleenew** [comment]
还没有合并吗？

**Comment by lucasjinreal** [comment]
@kylerwsm  Can u just merge this feature request? Or publish it to pub.dev?

**Comment by danieramiz** [comment]
dude, we need this. Show me what you got !

**Comment by lucasjinreal** [comment]
So the people who came back gone again.. I'd better fork this repo and maintain myself.

**Comment by lowdev** [comment]
Thank you about your PR @kylerwsm 
Awesome ^^
I'm really interesting by "leading widget" but we can't merge it because comments on other features.

I think we should create a PR for each new feature, It will be easier to review to test it and to merge it.
I can try to create a PR with only the "leading widget" feature. What do you think ?

Thank you again for your contribution

**Comment by kylerwsm** [comment]
> Thank you about your PR @kylerwsm
> Awesome ^^
> I'm really interesting by "leading widget" but we can't merge it because comments on other features.
> 
> I think we should create a PR for each new feature, It will be easier to review to test it and to merge it.
> I can try to create a PR with only the "leading widget" feature. What do you think ?
> 
> Thank you again for your contribution

Sounds good to me! Sorry that I don't have the time to work on this right now

**Comment by kylerwsm** [comment]
Hi all, I will be closing this PR as I do not have time to work on this right now. At the same time, I welcome anyone to pick this up.

File: lib/flappy_search_bar.dart
Comment (xBioDreadx) [reply lib/flappy_search_bar.dart]: That 80 height is actually a problem for me right now! Thanks for your solution! Lets hope changes will be merged soon...

**Pull request closed without merging**
